### ISSUE TITLE
Questions about embedding_model_path and phrase sampler settings for ProDVa

### ISSUE THREAD

**Issue by huggggoooooo** [sornkL/ProDVa#1]
Hi, thank you very much for releasing ProDVa — I’m trying to reproduce the protein design setting from the paper and use the released ProDVa-Molinst-SwissProtCLAP checkpoint on my evaluation set, which is a subset of Mol-Instruction protein-design dataset. I have two configuration questions about embedding_model_path and the phrase sampler.

**1. How should embedding_model_path be set for ProDVa-Molinst-SwissProtCLAP?**
Here is my chat.yaml:
```
model:
  model_name_or_path: "./ProDVa-Molinst-SwissProtCLAP"
  text_encoder_path: "./ProDVa-Molinst-SwissProtCLAP/text_tokenizer"
  language_model_path: "./ProDVa-Molinst-SwissProtCLAP/lm_tokenizer"
  phrase_encoder_path: "./ProDVa-Molinst-SwissProtCLAP/phrase_tokenizer"
  phrase_sampler_type: "PROTEIN_FRAGMENT"
  phrase_encoder_batch_size: 100000
  protein_fragment_mapping_file: "./Molinst-SwissProtCLAP/phrases.json"

infer:
  doc_top_k: 32
  embedding_model_path: "./pubmedbert-base-embeddings"  
  vector_store_path: "./CAMEO/index"
  do_sample: true
  max_new_tokens: 512
  temperature: 0.7
  top_k: 50
  protein_sequence_mapping_file: "./Molinst-SwissProtCLAP/training.json"
```

In the paper you mention using PubMedBERT to build faiss index, so I set embedding_model_path to a PubMedBERT embedding model downloading from https://huggingface.co/NeuML/pubmedbert-base-embeddings.
However, when I evaluate on Mol-Instructions–style tasks with this setup, while the generated sequences have high pae and plddt, their ProTrek scores are relatively low(about 1.4 in average while the groundtruth is 14.6), This makes me wonder:
- Is this the intended way to set embedding_model_path for ProDVa-Molinst-SwissProtCLAP?
- Should embedding_model_path point to a specific model/checkpoint (e.g. a particular PubMedBERT / sentence-transformers variant) that you used in your experiments?
- Or is there any other part of the infer config (e.g. doc_top_k, protein_sequence_mapping_file, etc.) that is crucial to match your original setup?
- Any guidance or example config would be very helpful.

**2. Phrase sampler settings in eval.yaml (N_TOKENS vs PROTEIN_FRAGMENT)**
In the example eval.yaml, the model block looks like:
```
model:
  model_name_or_path: "./ProDVa-Molinst-SwissProtCLAP"
  language_model_path: "./ProDVa-Molinst-SwissProtCLAP/lm_tokenizer"
  phrase_encoder_path: "./ProDVa-Molinst-SwissProtCLAP/phrase_tokenizer"
  phrase_sampler_type: "N_TOKENS"
  sampler_model_path: "./path/to/phrase_sampler_tokenizer"
  sampler_random_up: 12
  sampler_random_low: 8
  phrase_max_length: 5
  phrase_encoder_batch_size: 100000

```
Whereas for chat, I’m currently using:
```
phrase_sampler_type: "PROTEIN_FRAGMENT"
protein_fragment_mapping_file: "./Molinst-SwissProtCLAP/phrases.json"
```
I’m a bit confused about the intended usage here:
- What exactly is the role of phrase_sampler_type = "N_TOKENS" and sampler_model_path?
- Should sampler_model_path point to a text tokenizer (e.g., the same as text_encoder_path) or to a different model?
- For ProDVa, when doing protein design evaluation from natural language descriptions, should we be using phrase_sampler_type = "PROTEIN_FRAGMENT" (with phrases.json) consistently in both chat and eval configs?
- Is there any scenario where you recommend using "N_TOKENS" instead of "PROTEIN_FRAGMENT" for the released CAMEO model?

Right now my understanding is that ProDVa is supposed to use InterPro-derived fragments as dynamic “super-tokens” (i.e., PROTEIN_FRAGMENT with phrases.json), so I’d like to confirm the recommended settings and avoid misconfiguring the phrase sampler.

Additionally, in eval.yaml the argument test_data_file is described as a .txt file. For the PROTEIN_DESIGN task, what concrete format do you expect for this file? A minimal example of a valid test_data_file would be extremely helpful to make sure I’m evaluating the model correctly.

Thanks again for the great work and for releasing the code and models! Any clarification on these configuration details (especially a “canonical” chat.yaml / eval.yaml example and a sample test_data_file) would be greatly appreciated.

**Comment by sornkL** [comment]
Hi, thanks for your questions.

For the first question, we used the [microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext](https://huggingface.co/microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext) as the Embedding Model for retrieval. However, some other embedding models may also work. The key point here is to find the most relevant $K$ descriptions (as accurate as possible).

One important setting is to choose the appropriate database (aka supporting documents) during inference. In our experiments, for fair evaluations (e.g., to avoid data leakage), we choose the training set of each task as the database. In your setting, choosing the training set of Mol-Instructions as the database (`vector_store_path`) would produce much better performance. The reason is that the textual descriptions in the Mol-Instructions protein design task are much more complex than those in the CAMEO subset (with simple function keywords). Choosing the CAMEO subset as database may fail to retrieve proteins with relevant functions. Also remember to set the `protein_sequence_mapping_file` properly so that it matches the proteins in your database. Overall, I think these two steps (1. choosing the database and 2. set the mapping file properly) would help during the evaluation.

For the second question, I am sorry that I forgot to modify the config file when copying from the DVAGen framework. Thank you for pointing out the issue and I will fix the config file later! Coming back to your question, your previous setting used in the chat mode is correct (using `PROTEIN_FRAGMENT` as the `phrase_sampler_type` and the other sampler-based settings could be ignored since they are related to other `phrase_sampler_type` such as `N_TOKENS`).

I will briefly explain the differences between `N_TOKENS` and `PROTEIN_FRAGMENT` here. For more details of other Phrase Samplers, please refer to our [DVAGen paper](https://aclanthology.org/2025.emnlp-demos.26.pdf). `N_TOKENS` is used to split a tokenized text into random combinations of phrases (aka fragments) with N tokens length and tokens. It works in NLP tasks (e.g., language modeling) but fails in our preliminary experiments for protein design tasks. Therefore, we introduce `PROTEIN_FRAGMENTS` where biologically meaningful fragments are used in place of random N-gram fragments.

In conclusion, if you want to use ProDVa for protein design tasks, the use of `N_TOKENS` is consistently NOT recommended.

For the third question, thank you once again for pointing out the issue! The correct format should be a json file, which we have already released the test set used in our experiments: [CAMEO test set](https://huggingface.co/datasets/nwliu/CAMEO/blob/main/test.json) and [Mol-Instructions test set](https://huggingface.co/datasets/nwliu/Molinst-SwissProtCLAP/blob/main/test.json).

**Comment by sornkL** [comment]
I have updated the `chat` and `eval` config files and hope this helps!

**Comment by huggggoooooo** [comment]
Thanks again for your detailed reply!

I wanted to follow up with one technical observation about the Molinst-SwissProtCLAP FAISS index and one question about evaluation.

### 1. Embedding model and released FAISS index mismatch for Molinst-SwissProtCLAP

Following your suggestion, I switched to using the recommended embedding model download from https://huggingface.co/NeuML/pubmedbert-base-embeddings for retrieval, together with the released Molinst-SwissProtCLAP FAISS index from: https://huggingface.co/datasets/nwliu/Molinst-SwissProtCLAP.

However, I ran into what looks like a mismatch between the released `index` and the embeddings produced by this model.

I confirmed that the FAISS docstore is built from `Molinst-SwissProtCLAP/training.json`,  For example, the following instruction appears as the first entries in `training.json`:

> Plays a role in virus cell tropism, and may be required for efficient virus replication in macrophages. Belongs to the asfivirus MGF 100 family.

To debug the retrieval behavior, I compared:

1. The **stored embedding vector** from the released FAISS index for this instruction, and  
2. The **new embedding vector** computed on-the-fly for exactly the same text using  
   `microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext`  

For this instruction, I obtained:
```
[SHAPE]  
  stored_vec.shape: (768,)  
  new_vec.shape:    (768,)

[NORM]  
  ||stored_vec||: 1.0  
  ||new_vec||:    1.0

[SIMILARITY]  
  dot(stored, new): 0.1628  
  cosine(sim):       0.1628  
  L2 distance:       1.2940
```

So although the dimension and norms match, the cosine similarity is only ~0.16 and the L2 distance is relatively large. This suggests that the embeddings stored in the released `index` were possibly computed with a different embedding pipeline than the one obtained by directly using `BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext` in the way you described.

In practice, this mismatch caused issues that a similarity search using the current BiomedBERT encoder would not retrieve the corresponding document among the top-K neighbors. As a result, the retrieved descriptions were often unrelated, and the downstream designed sequences had poor ProTrek scores or clearly off-target behavior.

After I rebuilt the FAISS index from `Molinst-SwissProtCLAP/training.json` using exactly `microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext` as the embedding model the behavior became much more reasonable, the model’s design performance on my Mol-Instructions–style evaluation set looks much more sensible, ProTrek scores and structural metrics no longer collapse.

This makes me suspect that the Molinst-SwissProtCLAP `index` currently on Hugging Face may have been built with a slightly different embedding model or configuration than the one you recommended.

Would you mind confirming:
- Which exact embedding model / code path was originally used to build the released Molinst-SwissProtCLAP FAISS index?
- And whether you would recommend users to (re)build the index themselves using BiomedBERT, as I did, to match your intended setup?

I’m happy to share the small debug script I used (to reconstruct vectors from FAISS and compare with new embeddings) if that would be helpful.

---

### 2. Question about evaluation

I also noticed the following log when trying to use `dvagen infer/eval`:
[INFO] Evaluation on protein design metrics will be supported in a future version.  
      For now, please refer to https://github.com/PDFBench/PDFBench.

Since protein design evaluation is not fully supported in the current `dvagen eval` flow, my current workaround is:

- I run `dvagen chat --config_path chat_Molinst.yaml` in a separate process.
- Then I feed a list of descriptions to this chat process, one by one, treating it as a multi-turn conversation and parsing the generated sequences from the `Assistant:` lines.

This works reasonably well, but I’m not sure how close it is to your evaluation setup in the paper, and I am wondering whether using such a multi-turn chat setup (instead of evaluating each instruction independently with a fresh context) could affect the results compared to a fully stateless evaluation.

If you have a preferred / canonical way to run batched evaluation, I’d really appreciate any pointers.


Additionally, regarding **EvoLLaMA score** in PDFBench: as far as I understand, this metric also relies on an embedding model to compute the similarity between EvoLLaMA-generated descriptions and the ground-truth descriptions. If it’s convenient, could you share which specific embedding model / checkpoint you used there for computing the embeddings? Knowing this would help me align my evaluation setup more closely with yours.

---

Thanks again for your time and for making ProDVa and the associated resources available！ they’ve been very insightful to work with.

**Comment by sornkL** [comment]
Thank you for your questions!

For embedding models, I am sorry for the misunderstanding. We used the [NeuML/pubmedbert-base-embeddings](https://huggingface.co/NeuML/pubmedbert-base-embeddings) as the embedding model which was initially fine-tuned on the [microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext](https://huggingface.co/microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext). In our paper, we consistently used the former one (i.e., NeuML/pubmedbert-base-embeddings). However, switching to other models (e.g., microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext) may also work (but requires rebuilding the FAISS index).

Note that you should ALWAYS employ **the same embedding model** for both building the FAISS index and retrieval. The gap you mentioned is possibly caused by the fact that using different embedding models for these two steps. If you want to use our pre-constructed FAISS index on HuggingFace, please use the [NeuML/pubmedbert-base-embeddings](https://huggingface.co/NeuML/pubmedbert-base-embeddings) as the embedding model. Otherwise, employing the same embedding model for these two steps is critical.

BTW, I recommend (re)building the FAISS index in two cases:
- When switching to a different embedding model.
- When working with a new set of text–protein pair data.

---

For evaluation, running `dvagen chat` leads to a **single-turn conversation** (note that multi-turn conversation is not supported). Therefore, your current pipeline (i.e., running the command one by one) is completely reasonable. However, manually prompting ProDVa through the `dvagen chat` command for evaluation purposes is not recommend (manually copy-and-paste for each sample costs too much!!). Therefore, a straightforward approach is to run `dvagen eval`, which was originally designed for a two‑stage pipeline: first for inference (protein design) and then for evaluation (metric scoring). Due to the extensive effort required to migrate the evaluation codes, only the first stage is currently supported. The results will be saved as a json file which you could use it for further evaluation. For the second stage, please follow the instructions provided in our concurrent work, [PDFBench](https://github.com/PDFBench/PDFBench).

The weights of EvoLlama (i.e., the fine-tuned one used for evaluating Mol-Instructions task) can be found [here](https://huggingface.co/nwliu/EvoLlama-Oracle-Molinst-Protein-Design). Please follow these instructions:

- https://huggingface.co/nwliu/EvoLlama
- https://github.com/pdfbench/PDFBench?tab=readme-ov-file#31-protrek-and-evollama

**Comment by huggggoooooo** [comment]
Thank you very much for the detailed clarification!

That completely resolves my confusion. I don’t have any further questions on this, so I’m happy to consider the issue resolved. I’ll go ahead and close it. Thanks again for your time and for providing such clear guidance!
